slasher/db/kv/spanner_new.go

```diff
@@ -0,0 +1,110 @@
+package kv
+
+import (
+	"context"
+	"errors"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/slasher/detection/attestations/types"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+var errWrongSize = errors.New("wrong data length for min max span byte array")
+var highestObservedValidatorIdx uint64
+
+// GetValidatorSpan unmarshal a span from an encoded, flattened array.
+func (db *Store) GetValidatorSpan(ctx context.Context, spans []byte, validatorIdx uint64) (types.Span, error) {
+	ctx, span := trace.StartSpan(ctx, "slasherDB.getValidatorSpan")
+	defer span.End()
+
+	r := types.Span{}
+	if len(spans)%spannerEncodedLength != 0 {
+		return r, errWrongSize
+	}
+	origLength := uint64(len(spans)) / spannerEncodedLength
+	requestedLength := validatorIdx + 1
+	if origLength < requestedLength {
+		return r, nil
+	}
+	cursor := validatorIdx * spannerEncodedLength
+	r.MinSpan = bytesutil.FromBytes2(spans[cursor : cursor+2])
+	r.MaxSpan = bytesutil.FromBytes2(spans[cursor+2 : cursor+4])
+	sigB := [2]byte{}
+	copy(sigB[:], spans[cursor+4:cursor+6])
+	r.SigBytes = sigB
+	r.HasAttested = bytesutil.ToBool(spans[cursor+6])
+	return r, nil
+}
+
+// SetValidatorSpan marshal a validator span into an encoded, flattened array.
+func (db *Store) SetValidatorSpan(ctx context.Context, spans []byte, validatorIdx uint64, newSpan types.Span) ([]byte, error) {
```

Comment:
Does this need to be part a function receiving the DB? What could be a better function signature for it? I think the name could be improved a bit as well

Reply:
in order to implement the interface i have to make it a receiver function. will be happy to remove this requirement

Reply:
fixed it as you suggested

Reply:
thanks